internal/topology/variables/cluster_variable_defaulting.go

```diff
@@ -86,83 +81,143 @@ func defaultClusterVariables(clusterVariables []clusterv1.ClusterVariable, clust
-			fmt.Sprintf("invalid schema in ClusterClass for variable %q: error to convert schema %v", clusterClassVariable.Name, errs))}
+			fmt.Sprintf("invalid schema in ClusterClass for variable %q: error to convert schema %v", definition.Name, errs))}
 	}
 
 	var value interface{}
 	// If the variable already exists, parse the current value.
-	if clusterVariable != nil && len(clusterVariable.Value.Raw) > 0 {
-		if err := json.Unmarshal(clusterVariable.Value.Raw, &value); err != nil {
+	if currentValue != nil && len(currentValue.Value.Raw) > 0 {
+		if err := json.Unmarshal(currentValue.Value.Raw, &value); err != nil {
 			return nil, field.ErrorList{field.Invalid(fldPath, "",
-				fmt.Sprintf("failed to unmarshal variable value %q: %v", string(clusterVariable.Value.Raw), err))}
+				fmt.Sprintf("failed to unmarshal variable %q value %q: %v", definition.Name, string(currentValue.Value.Raw), err))}
 		}
 	}
 
 	// Structural schema defaulting does not work with scalar values,
 	// so we wrap the schema and the variable in objects.
 	// <variable-name>: <variable-value>
 	wrappedVariable := map[string]interface{}{
-		clusterClassVariable.Name: value,
+		definition.Name: value,
 	}
 	// type: object
 	// properties:
 	//   <variable-name>: <variable-schema>
 	wrappedSchema := &apiextensions.JSONSchemaProps{
 		Type: "object",
 		Properties: map[string]apiextensions.JSONSchemaProps{
-			clusterClassVariable.Name: *apiExtensionsSchema,
+			definition.Name: *apiExtensionsSchema,
 		},
 	}
 
 	// Default the variable via the structural schema library.
 	ss, err := structuralschema.NewStructural(wrappedSchema)
 	if err != nil {
 		return nil, field.ErrorList{field.Invalid(fldPath, "",
-			fmt.Sprintf("failed defaulting variable %q: %v", clusterVariable.Name, err))}
+			fmt.Sprintf("failed defaulting variable %q: %v", currentValue.Name, err))}
 	}
 	structuraldefaulting.Default(wrappedVariable, ss)
 
 	// Marshal the defaulted value.
-	defaultedVariableValue, err := json.Marshal(wrappedVariable[clusterClassVariable.Name])
+	defaultedVariableValue, err := json.Marshal(wrappedVariable[definition.Name])
 	if err != nil {
 		return nil, field.ErrorList{field.Invalid(fldPath, "",
-			fmt.Sprintf("failed to marshal default value of variable %q: %v", clusterClassVariable.Name, err))}
+			fmt.Sprintf("failed to marshal default value of variable %q: %v", definition.Name, err))}
 	}
-	return &clusterv1.ClusterVariable{
-		Name: clusterClassVariable.Name,
+	v := &clusterv1.ClusterVariable{
+		Name: definition.Name,
 		Value: apiextensionsv1.JSON{
 			Raw: defaultedVariableValue,
 		},
-	}, nil
+		DefinitionFrom: definition.From,
+	}
+
+	return v, nil
+}
+
+// getAllVariables returns a correctly ordered list of all variables defined in the ClusterClass and the Cluster.
+func getAllVariables(values []clusterv1.ClusterVariable, valuesIndex map[string]map[string]clusterv1.ClusterVariable, definitions []clusterv1.ClusterClassStatusVariable) []clusterv1.ClusterVariable {
```

Comment:
I think it would be good to have a unit test at this level to cover all the permutations